internal/batch/batch_processor.go

```diff
@@ -101,103 +130,249 @@ func (bw *batchWork) estimateSize() int64 {
+	added := false
+	skip := false
+	// Check it's not in the recently flushed lish
+	for _, flushedSequence := range bp.flushedSequences {
+		if newWork.msg.Sequence == flushedSequence {
+			log.L(bp.ctx).Debugf("Ignoring add of recently flushed message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+			skip = true
+			break
+		}
+	}
+	// Build the new sorted work list, checking there for duplicates too
+	for _, work := range bp.assemblyQueue {
+		if newWork.msg.Sequence == work.msg.Sequence {
+			log.L(bp.ctx).Debugf("Ignoring duplicate add of message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+			skip = true
+		}
+		if !added && !skip && newWork.msg.Sequence < work.msg.Sequence {
+			newQueue = append(newQueue, newWork)
+			added = true
+		}
+		newQueue = append(newQueue, work)
+	}
+	if !added && !skip {
+		newQueue = append(newQueue, newWork)
+		added = true
+	}
+	if added {
+		log.L(bp.ctx).Debugf("Added message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+		bp.assemblyQueueBytes += newWork.estimateSize()
+	}
+	bp.assemblyQueue = newQueue
+	full = len(bp.assemblyQueue) >= int(bp.conf.BatchMaxSize) || (bp.assemblyQueueBytes >= bp.conf.BatchMaxBytes)
+	overflow = len(bp.assemblyQueue) > 1 && (bp.assemblyQueueBytes > bp.conf.BatchMaxBytes)
+	return full, overflow
+}
+
+func (bp *batchProcessor) startFlush(overflow bool) (id *fftypes.UUID, flushAssembly []*batchWork, byteSize int64) {
+	bp.statusMux.Lock()
+	defer bp.statusMux.Unlock()
+	// Star the clock
+	bp.flushStatus.Blocked = false
+	bp.flushStatus.LastFlushTime = fftypes.Now()
+	// Split the current work if required for overflow
+	overflowWork := make([]*batchWork, 0)
+	if overflow {
+		lastElem := len(bp.assemblyQueue) - 1
+		flushAssembly = append(flushAssembly, bp.assemblyQueue[:lastElem]...)
+		overflowWork = append(overflowWork, bp.assemblyQueue[lastElem])
+	} else {
+		flushAssembly = bp.assemblyQueue
+	}
+	// We need to keep track of the sequences we're flushing, because until we finish our flush
+	// the batch processor might be re-queuing the same messages to use due to rewinds.
+	// We keep all of the last batch, and up to twice the batch size over time (noting our channel
+	// size is our batch size - so the batch manager cannot get further than that ahead).
+	newFlushedSeqLen := len(flushAssembly) + len(bp.flushedSequences)
+	maxFlushedSeqLen := int(2 * bp.conf.BatchMaxSize)
+	if newFlushedSeqLen > maxFlushedSeqLen && maxFlushedSeqLen > len(bp.flushedSequences) {
+		newFlushedSeqLen = maxFlushedSeqLen
+	}
+	newFlushedSequnces := make([]int64, newFlushedSeqLen)
+	for i, fs := range flushAssembly {
+		newFlushedSequnces[i] = fs.msg.Sequence
+	}
+	for i := 0; i < newFlushedSeqLen-len(flushAssembly); i++ {
+		newFlushedSequnces[i+len(flushAssembly)] = bp.flushedSequences[i]
```

Comment:
Are we intentionally keeping the _first_ elements in `bp.flushedSequences` and dropping ones off the end? Or do we actually want to preserve the tail and drop the earliest elements?

Reply:
Also if you initialize the slice with `make([]int64, len(flushAssembly), newFlushedSeqLen)`, then I think maybe you could get fancy here by replacing the loop with an `append`... not sure what ultimately reads cleaner though.

Reply:
Because the existing `flushedSequences` might be a bunch of small chunks, the right change here seemed to be to reverse the order of the array elements as we add them (rather than as we remove them).

That way the chunks are ordered newest to oldest, and the elements are ordered newest to oldest.